component/loki/source/kubernetes/kubetail/tailer.go

```diff
@@ -0,0 +1,323 @@
+		// be returned alongside an EOF.
+		if len(line) != 0 {
+			entryTimestamp, entryLine := parseKubernetesLog(line)
+			if !entryTimestamp.After(lastReadTime) {
+				continue
+			}
+			lastReadTime = entryTimestamp
+
+			entry := loki.Entry{
+				Labels: t.lset,
+				Entry: logproto.Entry{
+					Timestamp: entryTimestamp,
+					Line:      entryLine,
+				},
+			}
+
+			select {
+			case <-ctx.Done():
+				return nil
+			case ch <- entry:
+				// Save position after it's been sent over the channel.
+				t.opts.Positions.Put(positionsEnt.Path, positionsEnt.Labels, entryTimestamp.UnixMicro())
+				t.target.Report(entryTimestamp, nil)
+			}
+		}
+
+		// Return an error if our stream closed. The caller will reopen the tailer
+		// forever until our tailer is closed.
+		//
+		// Even if EOF is returned, we still want to allow the tailer to retry
+		// until the tailer is shutdown; EOF being returned doesn't necessarily
+		// indicate that the logs are done, and could point to a brief network
+		// outage.
+		if err != nil && errors.Is(err, io.EOF) {
+			return nil
+		} else if err != nil {
+			return err
+		}
+	}
+}
+
+// containerTerminated determines whether the container this tailer was
+// watching has terminated and won't restart. If containerTerminated returns
+// true, it means that no more logs will appear for the watched target.
+func (t *tailer) containerTerminated(ctx context.Context) (terminated bool, err error) {
+	var (
+		key           = t.target.NamespacedName()
+		containerName = t.target.ContainerName()
+	)
+
+	podInfo, err := t.opts.Client.CoreV1().Pods(key.Namespace).Get(ctx, key.Name, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	// The pod UID is different than the one we were tailing; our UID has
+	// terminated.
+	if podInfo.GetUID() != kubetypes.UID(t.target.UID()) {
+		return true, nil
+	}
+
+	containerInfo, containerType, found := findContainerStatus(podInfo, containerName)
+	if !found {
+		return false, fmt.Errorf("could not find container %q in pod status", containerName)
+	}
+
+	restartPolicy := podInfo.Spec.RestartPolicy
+
+	switch containerType {
+	case containerTypeApp:
+		// An app container will only restart if:
+		//
+		// * It is in a waiting (meaning it's waiting to run) or running state
+		//   (meaning it already restarted before we had a chance to check)
+		// * It terminated with any exit code and restartPolicy is Always
+		// * It terminated with non-zero exit code and restartPolicy is not Never
+		//
+		// https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy
+		switch {
+		case containerInfo.State.Waiting != nil || containerInfo.State.Running != nil:
+			return false, nil // Container will restart
+		case restartPolicy == corev1.RestartPolicyAlways:
+			return false, nil // Container will restart
+		case containerInfo.State.Terminated != nil && restartPolicy == corev1.RestartPolicyOnFailure && containerInfo.State.Terminated.ExitCode != 0:
```

Comment:
Do we want to check for `restartPolicy == OnFailure` or `restartPolicy != Never` like the comment says?